File: virtual-world-cpp/VirtualWorld/Animals/Turtle.cpp
#include <string>
#include "Turtle.h"

Turtle::Turtle(int x, int y, World *world) : Animal(x, y, 2, 1, 0, turtleCode, world) {}

bool Turtle::AttackPaired(Organism *attacker) {
    if(attacker->getStrength() >= 5)
        return false;
    this->world->AddMessage("Turtle blocked attack from " + attacker->GetName() + " at " + std::to_string(attacker->getX()) + " " + std::to_string(attacker->getY()) + "!");
    return true;
}

bool Turtle::TryToBreed(Organism *other) {
    if(dynamic_cast<Turtle*>(other) != nullptr)
        return true;
    return false;
}

std::string Turtle::GetName() {
    return "Turtle";
}

Organism *Turtle::Clone(int x, int y, World *world) {
    return new Turtle(x, y, world);
}

void Turtle::Action() {
    int random = rand() % 4;
    if(random == 0)
        Animal::Action();
}

File: virtual-world-java/src/main/java/me/knasuta/virtualworld/simulation/animals/Turtle.java
package me.knasuta.virtualworld.simulation.animals;

import me.knasuta.virtualworld.simulation.Organism;
import me.knasuta.virtualworld.simulation.Point;
import me.knasuta.virtualworld.simulation.World;

import java.awt.*;

public class Turtle extends Animal {

    public Turtle(Point location, World world) {
        super(location, 2, 1, 0, world);
    }

    @Override
    public boolean AttackPaired(Organism other) {
        if(other.getStrength() < 5){
            world.AddMessage("Turtle defended itself from " + other.getName() + " at " + this.location.toString());
            return true;
        }
        return false;
    }

    @Override
    public String getName() {
        return "Turtle";
    }

    @Override
    public Organism Clone(Point location, World world) {
        return new Turtle(location, world);
    }

    @Override
    public Color getColor() {
        return Color.GREEN;
    }

    @Override
    public void Action(){
        if(Math.random() < 0.25)
            super.Action();
    }
}
